Pick up photo from gallery for external sources ends up with an app crash

E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #4
                  Process: com.ibm.watson.developer_cloud.android.myapplication, PID: 2902
                  java.lang.RuntimeException: An error occurred while executing doInBackground()
                      at android.os.AsyncTask$3.done(AsyncTask.java:325)
                      at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
                      at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:242)
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                      at java.lang.Thread.run(Thread.java:761)
                   Caused by: java.lang.RuntimeException: java.io.FileNotFoundException: /external/images/media/34 (No such file or directory)
                      at com.ibm.watson.developer_cloud.service.WatsonService$1.execute(WatsonService.java:182)
                      at com.ibm.watson.developer_cloud.android.myapplication.MainActivity$RecognitionTask.doInBackground(MainActivity.java:103)
                      at com.ibm.watson.developer_cloud.android.myapplication.MainActivity$RecognitionTask.doInBackground(MainActivity.java:96)
                      at android.os.AsyncTask$2.call(AsyncTask.java:305)
                      at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243) 
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                      at java.lang.Thread.run(Thread.java:761) 
                   Caused by: java.io.FileNotFoundException: /external/images/media/34 (No such file or directory)
                      at java.io.FileInputStream.open(Native Method)
                      at java.io.FileInputStream.<init>(FileInputStream.java:146)
                      at okio.Okio.source(Okio.java:168)
                      at okhttp3.RequestBody$3.writeTo(RequestBody.java:117)
                      at okhttp3.MultipartBody.writeOrCountBytes(MultipartBody.java:171)
                      at okhttp3.MultipartBody.writeTo(MultipartBody.java:113)
                      at okhttp3.internal.http.HttpEngine$NetworkInterceptorChain.proceed(HttpEngine.java:756)
                      at okhttp3.internal.http.HttpEngine.readResponse(HttpEngine.java:613)
                      at okhttp3.RealCall.getResponse(RealCall.java:244)
                      at okhttp3.RealCall$ApplicationInterceptorChain.proceed(RealCall.java:201)
                      at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:163)
                      at okhttp3.RealCall.execute(RealCall.java:57)
                      at com.ibm.watson.developer_cloud.service.WatsonService$1.execute(WatsonService.java:179)
                      at com.ibm.watson.developer_cloud.android.myapplication.MainActivity$RecognitionTask.doInBackground(MainActivity.java:103) 
                      at com.ibm.watson.developer_cloud.android.myapplication.MainActivity$RecognitionTask.doInBackground(MainActivity.java:96) 
                      at android.os.AsyncTask$2.call(AsyncTask.java:305) 
                      at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                      at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243) 
                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                      at java.lang.Thread.run(Thread.java:761) 
Disconnected from the target VM, address: 'localhost:8870', transport: 'socket'

--- a/example/build.gradle
+++ b/example/build.gradle
@@ -12,7 +12,7 @@ android {
   defaultConfig {
     applicationId "com.ibm.watson.developer_cloud.android.myapplication"
     testInstrumentationRunner "android.support.test.runner.AndroidJUnitRunner"
-    minSdkVersion 9
+    minSdkVersion 11
     targetSdkVersion 22
     versionCode 1
     versionName "1.0"

--- a/library/build.gradle
+++ b/library/build.gradle
@@ -15,7 +15,7 @@ android {
   buildToolsVersion "23.0.2"
 
   defaultConfig {
-    minSdkVersion 8
+    minSdkVersion 11
     targetSdkVersion 23
     versionCode 1
     versionName "1.0"

--- a/library/src/main/java/com/ibm/watson/developer_cloud/android/library/camera/GalleryHelper.java
+++ b/library/src/main/java/com/ibm/watson/developer_cloud/android/library/camera/GalleryHelper.java
@@ -2,10 +2,12 @@ package com.ibm.watson.developer_cloud.android.library.camera;
 
 import android.app.Activity;
 import android.content.Intent;
+import android.database.Cursor;
 import android.graphics.Bitmap;
 import android.graphics.BitmapFactory;
 import android.net.Uri;
 import android.provider.MediaStore;
+import android.support.v4.content.CursorLoader;
 import android.util.Log;
 import java.io.File;
 import java.io.FileNotFoundException;
@@ -46,7 +48,8 @@ public final class GalleryHelper {
   public File getFile(int resultCode, Intent data) {
     if(resultCode == activity.RESULT_OK) {
       Uri targetUri = data.getData();
-      return new File(targetUri.getPath());
+      //return new File(targetUri.getPath());
+      return new File(getRealPathFromURI(targetUri));
     }
     Log.e(TAG, "Result Code was not OK");
     return null;
@@ -72,5 +75,20 @@ public final class GalleryHelper {
     }
     Log.e(TAG, "Result Code was not OK");
     return null;
+  }
+
+  /**
+   * This method returns a path, given a content URI
+   *
+   * @param contentUri URI of some image
+   * @return A string path
+   */
+  public String getRealPathFromURI(Uri contentUri) {
+    String[] proj = { MediaStore.Images.Media.DATA };
+    CursorLoader cLoader = new CursorLoader(activity.getApplicationContext(), contentUri, proj, null, null, null);
+    Cursor cursor = cLoader.loadInBackground();
+    int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
+    cursor.moveToFirst();
+    return cursor.getString(column_index);
   }
 }

--- a/library/src/main/java/com/ibm/watson/developer_cloud/android/library/camera/GalleryHelper.java
+++ b/library/src/main/java/com/ibm/watson/developer_cloud/android/library/camera/GalleryHelper.java
@@ -48,7 +48,6 @@ public final class GalleryHelper {
   public File getFile(int resultCode, Intent data) {
     if(resultCode == activity.RESULT_OK) {
       Uri targetUri = data.getData();
-      //return new File(targetUri.getPath());
       return new File(getRealPathFromURI(targetUri));
     }
     Log.e(TAG, "Result Code was not OK");

